pkg/backend/filestate/lock.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filestate
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/pulumi/pulumi/pkg/backend"
+	"github.com/pulumi/pulumi/pkg/util/logging"
+	"os"
+	"os/user"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+type lockContent struct {
+	Pid       int       `json:"pid"`
+	Username  string    `json:"username"`
+	Hostname  string    `json:"hostname"`
+	Timestamp time.Time `json:"timestamp"`
+}
+
+func newLockContent() (*lockContent, error) {
+	u, err := user.Current()
+	if err != nil {
+		return nil, err
+	}
+	hostname, err := os.Hostname()
+	if err != nil {
+		return nil, err
+	}
+	return &lockContent{
+		Pid:       os.Getpid(),
+		Username:  u.Username,
+		Hostname:  hostname,
+		Timestamp: time.Now(),
+	}, nil
+}
+
+func (l *lockContent) String() string {
+	return fmt.Sprintf("%v@%v (pid %v)", l.Username, l.Hostname, l.Pid)
+}
+
+func (b *localBackend) checkForLock(ctxt context.Context, stackRef backend.StackReference) error {
+	allFiles, err := listBucket(b.bucket, b.LockDir())
+	if err != nil {
+		return err
+	}
+
+	var lockKeys []string
+	for i := len(allFiles) - 1; i >= 0; i-- {
+		if allFiles[i].IsDir {
+			continue
+		}
+		fileKey := allFiles[i].Key
+		if strings.HasPrefix(fileKey, b.lockPrefix(stackRef.Name())) {
+			lockKeys = append(lockKeys, fileKey)
+		}
+	}
+
+	if len(lockKeys) > 0 {
+		errorString := fmt.Sprintf("the stack is current locked by %v lock(s). Either wait for the other "+
+			"process(es) to end or manually delete the lock file(s).", len(lockKeys))
+
+		for _, lock := range lockKeys {
+			content, err := b.bucket.ReadAll(ctxt, lock)
+			if err != nil {
+				return err
+			}
+			l := &lockContent{}
+			err = json.Unmarshal(content, &l)
+			if err != nil {
+				return err
+			}
+
+			// this is kind of weird but necessary because url is a string and not a url.URL
+			url := b.url + filepath.Join("/", lock)
+			errorString = errorString + fmt.Sprintf("\n  %v: created by %v", url, l.String())
+		}
+
+		return fmt.Errorf(errorString)
+	}
+	return nil
+}
+
+func (b *localBackend) checkForLockRace(stackRef backend.StackReference) error {
+	// Check the locks to make sure ONLY our lock exists. If we find multiple locks then we know
+	// we had a race condition and we should clean up and abort
```

Comment:
i'm not certain at all what sort of global consistency views the cloud providers ive for buckets.  Is it the case that if you write a file, and then list that you will see the writes from another client?  for example, say two clients do the following:

```
step 1: write lock file.
step 2: list bucket
```

Do we feel that it's not possible for the cloud-provider/clients to show both clients that they only see the lock file written by that client?

Note: it may be hte case that this is a race, but htat's narrow enough to be ok.  If the purpose is just trying to lock for the long periods of time when an update it happening, but not strictly trying to be totally race free.

Reply:
I haven't done my own research but was basing this based off of these two comments. I believe that the strongly consistent read-after-write means we are safe here. Azure and GCS are strongly consistent.

https://github.com/pulumi/pulumi/pull/2455#issuecomment-465114882

https://github.com/pulumi/pulumi/pull/2455#issuecomment-465447180

Reply:
sounds good to me.